### PR TITLE
[InstCombine] Improve handling of `not` and free inversion.

### PR DESCRIPTION
**Pull request by goldsteinn**
- [InstCombine] Add tests for folding `(X + Y) - (W + Z)`; NFC
- [InstCombine] Add folds for `(X + Y) - (W + Z)`
- [InstCombine] Make `isFreeToInvert` check recursively.
- [InstCombine] Add additional tests for free inversion; NFC
- [InstCombine] add `getFreeInverted` to perform folds for free inversion of op
- [InstCombine] Replace `isFreeToInvert` + `CreateNot` with `getFreelyInverted`
- [InstCombine] Add transform for `~X + ~Y)` -> `-2 - Y - X`

### CODE REVIEW

**Comment by RKSimon** [comment]
I'm curious how you intend to commit this stack of PRs? Are you intending to squash + merge?

**Comment by goldsteinn** [comment]
> I'm curious how you intend to commit this stack of PRs? Are you intending to squash + merge?

I was/am intending to push each commit discretely.
As they are approved I generally create patches, apply to main, then push.
Not the best workflow for PRs?
Is there a better way to create PR with many independent commits?

**Comment by RKSimon** [comment]
I'm still trying to work out how best to create stacked PRs (especially when adding extra tests and want to show the diff) - everyone seems to have their own opinion on them right now :)

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 1]
```diff
@@ -245,32 +246,38 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
     if (match(V, PatternMatch::m_AnyIntegralConstant()))
       return true;
 
+    if (Depth++ >= MaxAnalysisRecursionDepth)
```
Comment (RKSimon) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:251]: (style) For recursion methods add Depth on the recursive calls themselves (isFreeToInvert(X,Y, Depth + 1)

Comment (goldsteinn) [reply llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:251]: This is something there is no consistency across LLVM with. I know in SelectionDAG we do `Depth; Depth + 1` but somewhere like ValueTracking we use `Depth++; Depth`.

Personally I like the `Depth++` way b.c you don't have to type `Depth + 1` everywhere (easier to not mess up imo), and it seems more consistent with what a see elsewhere in InstCombine / the middle end.

Unless you feel strongly would prefer to keep as is.

**Comment by goldsteinn** [comment]
> I'm still trying to work out how best to create stacked PRs (especially when adding extra tests and want to show the diff) - everyone seems to have their own opinion on them right now :)

I don't really know either. If you know a better way than this I'm all ears.

**Comment by goldsteinn** [comment]
> > I'm still trying to work out how best to create stacked PRs (especially when adding extra tests and want to show the diff) - everyone seems to have their own opinion on them right now :)
> 
> I don't really know either. If you know a better way than this I'm all ears.

What might improve the situation is if we could create branch on the llvm-project repo. I.e `users/<gh-username>/<branch-name>`. Then we could create series of PRs i.e
`main` <- `users/<user>/feature-patch-0` <- `users/<user>/feature-patch-1` <-... `users/<user>/feature-patch-N`

**Comment by goldsteinn** [comment]
ping.

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 2]
```diff
@@ -230,49 +230,114 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
+    // Selects/min/max with invertible operands are freely invertible
+    if (IsSelect || match(V, m_MaxOrMin(PatternMatch::m_Value(A),
+                                        PatternMatch::m_Value(B)))) {
+      if (!getFreeInverted(A, A->hasOneUse(), /*Builder*/ nullptr, Depth) ||
+          !getFreeInverted(B, B->hasOneUse(), /*Builder*/ nullptr, Depth))
+        return nullptr;
+      if (Builder != nullptr) {
+        A = getFreeInverted(A, A->hasOneUse(), Builder, Depth);
+        B = getFreeInverted(B, B->hasOneUse(), Builder, Depth);
+        assert(A != nullptr && B != nullptr);
```
Comment (RKSimon) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:321]: (style) missing assert message

**Comment by goldsteinn** [comment]
ping.

**Comment by goldsteinn** [comment]
ping.

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 3]
```diff
@@ -248,32 +249,38 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
+             isFreeToInvert(B, B->hasOneUse(), Depth);
+
+    // If `V` is of the form `A - B` then `-1 - V` can be folded into
+    // `~A + B` if we are willing to invert all of the uses.
+    if (match(V, m_Sub(PatternMatch::m_Value(A), PatternMatch::m_Value())))
+      return isFreeToInvert(A, A->hasOneUse(), Depth);
+
+    // Selects/min/max with invertible operands are freely invertible
+    if (match(V, m_Select(PatternMatch::m_Value(), PatternMatch::m_Value(A),
+                          PatternMatch::m_Value(B))) ||
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h]: With this generalization, the select case should check `shouldAvoidAbsorbingNotIntoSelect()` to avoid inverting logical and/or into non-canonical form.

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 4]
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
-  ///
-  /// See also: canFreelyInvertAllUsersOf()
-  static bool isFreeToInvert(Value *V, bool WillInvertAllUses,
-                             unsigned Depth = 0) {
+  /// Return nonnull value if V is free to invert (with condition) regarding
+  /// WillInvertAllUses.
+  /// If Builder is nonnull, it will return a simplified ~V
+  /// If builder is null, it will return an arbitrary nonnull value (not
+  /// dereferenceable).
+  static Value *getFreeInverted(Value *V, bool WillInvertAllUses,
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:241]: I'd call this getFreelyInverted().

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 5]
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
-  /// is true, work under the assumption that the caller intends to remove all
-  /// uses of V and only keep uses of ~V.
-  ///
-  /// See also: canFreelyInvertAllUsersOf()
-  static bool isFreeToInvert(Value *V, bool WillInvertAllUses,
-                             unsigned Depth = 0) {
+  /// Return nonnull value if V is free to invert (with condition) regarding
+  /// WillInvertAllUses.
+  /// If Builder is nonnull, it will return a simplified ~V
+  /// If builder is null, it will return an arbitrary nonnull value (not
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:239]: builder -> Builder for consistency with previous sentence.

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 6]
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
+  /// dereferenceable).
+  static Value *getFreeInverted(Value *V, bool WillInvertAllUses,
+                                BuilderTy *Builder, unsigned Depth = 0) {
+    static Value *const kNonNull = reinterpret_cast<Value *>(uintptr_t(1));
     // ~(~(X)) -> X.
-    if (match(V, m_Not(PatternMatch::m_Value())))
-      return true;
+    Value *A, *B;
+    if (match(V, m_Not(PatternMatch::m_Value(A))))
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:246]: As you're touching all the code in here anyway, I'd add a `using namespace llvm::PatternMatch` inside this function, so make the implementation more idiomatic...

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 7]
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
-  static bool isFreeToInvert(Value *V, bool WillInvertAllUses,
-                             unsigned Depth = 0) {
+  /// Return nonnull value if V is free to invert (with condition) regarding
+  /// WillInvertAllUses.
+  /// If Builder is nonnull, it will return a simplified ~V
+  /// If builder is null, it will return an arbitrary nonnull value (not
+  /// dereferenceable).
+  static Value *getFreeInverted(Value *V, bool WillInvertAllUses,
+                                BuilderTy *Builder, unsigned Depth = 0) {
+    static Value *const kNonNull = reinterpret_cast<Value *>(uintptr_t(1));
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:243]: `k` prefix is not typical LLVM style...

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 8]
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
+    // If `V` is of the form `Constant - A` then `-1 - V` can be folded into
+    // `A + (-1 - Constant)` if we are willing to invert all of the uses.
+    if (match(V, m_Sub(PatternMatch::m_Value(A), PatternMatch::m_Value(B)))) {
+      if (auto *AV = getFreeInverted(A, A->hasOneUse(), Builder, Depth))
+        return Builder ? Builder->CreateAdd(AV, B) : kNonNull;
+      return nullptr;
+    }
+
+    // If `V` is of the form `Constant - A` then `-1 - V` can be folded into
+    // `A + (-1 - Constant)` if we are willing to invert all of the uses.
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:300]: Comment is about sub, implementation is for ashr.

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 9]
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
+    }
+
+    Value *Cond;
+    // LogicOps are special in that we canonicalize them at the cost of an
+    // instruction.
+    bool IsSelect =
+        match(V, m_Select(PatternMatch::m_Value(Cond), PatternMatch::m_Value(A),
+                          PatternMatch::m_Value(B))) &&
+        !match(V, PatternMatch::m_LogicalOp(PatternMatch::m_Value(),
+                                            PatternMatch::m_Value()));
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:314]: Okay, I see you ended up basically doing the same thing as shouldAvoidAbsorbingNotIntoSelect() in a following commit. Would still be good to use the standard helper for that.

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 10]
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
-        match(V,
-              m_MaxOrMin(PatternMatch::m_Value(A), PatternMatch::m_Value(B))))
-      return isFreeToInvert(A, A->hasOneUse(), Depth) &&
-             isFreeToInvert(B, B->hasOneUse(), Depth);
-
-    return false;
+    if (IsSelect || match(V, m_MaxOrMin(PatternMatch::m_Value(A),
+                                        PatternMatch::m_Value(B)))) {
+      if (!getFreeInverted(A, A->hasOneUse(), /*Builder*/ nullptr, Depth) ||
+          !getFreeInverted(B, B->hasOneUse(), /*Builder*/ nullptr, Depth))
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:319]: You could only check A without Builder upfront, and then check B with Builder, avoiding one duplicate check.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 11]
```diff
@@ -4410,6 +4410,11 @@ Instruction *InstCombinerImpl::foldNot(BinaryOperator &I) {
   if (Instruction *NewXor = foldNotXor(I, Builder))
     return NewXor;
 
+  // TODO: Could handle multi-use better by checking if all uses of NotOp (other
+  // than I) can be inverted.
+  if (Value *R = getFreeInverted(NotOp, NotOp->hasOneUse(), &Builder))
+    return replaceInstUsesWith(I, R);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4453]: There is a `sinkNotIntoXor()` fold that I think gets obsoleted by this one? Any test changes if you drop it?

**Comment by nikic** [comment]
Only looked at the first three commits, which imho could be a separate PR (maybe drop the rest from this one for now).

I think after adding getFreeInverted() it would be good to also go through existing users of isFreeToInvert() and replace them with getFreeInverted() where this is easy, e.g. in foldICmpOrXX(). Directly creating the inverted op instead of letting it get inverted separately does seem a lot more robust to me.

**Comment by goldsteinn** [comment]
> Only looked at the first three commits, which imho could be a separate PR (maybe drop the rest from this one for now).
> 
> I think after adding getFreeInverted() it would be good to also go through existing users of isFreeToInvert() and replace them with getFreeInverted() where this is easy, e.g. in foldICmpOrXX(). Directly creating the inverted op instead of letting it get inverted separately does seem a lot more robust to me.

Thank you for the review.

Regarding replacing `getFreeInverted` with `isFreeToInvert` + (creating a `not`). So I do add a fold for using `getFreeInverted` in `not` fold. My feeling was we have places where we create a `not` in a canonicalizing effort and more places recognize `m_Not` than `isFreeToInvert`, so it made sense to let the new pattern be checked by other folds before stripping the `not`. But not married to that idea.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 11 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4453]: Some tests still rely on it, but did't poke around that deeply.

**Comment by goldsteinn** [comment]
> > Only looked at the first three commits, which imho could be a separate PR (maybe drop the rest from this one for now).
> > I think after adding getFreeInverted() it would be good to also go through existing users of isFreeToInvert() and replace them with getFreeInverted() where this is easy, e.g. in foldICmpOrXX(). Directly creating the inverted op instead of letting it get inverted separately does seem a lot more robust to me.
> 
> Thank you for the review.
> 
> Regarding replacing `getFreeInverted` with `isFreeToInvert` + (creating a `not`). So I do add a fold for using `getFreeInverted` in `not` fold. My feeling was we have places where we create a `not` in a canonicalizing effort and more places recognize `m_Not` than `isFreeToInvert`, so it made sense to let the new pattern be checked by other folds before stripping the `not`. But not married to that idea.

Okay, its split to 4 four commits related to the free inversion (re-titled).
The fourth commit is essentially NFC (although not quite due to change in inst order/names) and puts `getFreelyInverted` in a few places.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 12]
```diff
@@ -4317,13 +4327,15 @@ Instruction *InstCombinerImpl::foldNot(BinaryOperator &I) {
...
-      return replaceInstUsesWith(I, InvMaxMin);
+        isFreeToInvert(X, X->hasOneUse())) {
+      if (Value *NotY = getFreelyInverted(Y, Y->hasOneUse(), &Builder)) {
+        Intrinsic::ID InvID = getInverseMinMaxIntrinsic(II->getIntrinsicID());
+        Value *NotX = getFreelyInverted(X, X->hasOneUse(), &Builder);
+        assert(NotX != nullptr &&
+               "isFreeToInvert desynced with getFreelyInverted");
+        Value *InvMaxMin = Builder.CreateBinaryIntrinsic(InvID, NotX, NotY);
+        return replaceInstUsesWith(I, InvMaxMin);
+      }
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4338]: I think this implements the same logic as getFreelyInverted() itself -- can we just call getFreelyInverted() on II?

(More generally, we should probably have a generic fold calling getFreelyInverted() here, which likely subsumes some of the other folds in this function. But maybe that's for a followup.)

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 13]
```diff
@@ -2198,12 +2198,14 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
...
-    Value *NotOp1 = Builder.CreateNot(Op1);
-    return BinaryOperator::CreateSub(NotOp1, NotOp0);
+  if (!match(Op0, m_Sub(m_ImmConstant(), m_Value())) &&
+      isFreeToInvert(Op0, Op0->hasOneUse())) {
+    if (Value *NotOp1 = getFreelyInverted(Op1, Op1->hasOneUse(), &Builder)) {
+      Value *NotOp0 = getFreelyInverted(Op0, Op0->hasOneUse(), &Builder);
+      assert(NotOp0 != nullptr &&
+             "isFreeToInvert desynced with getFreelyInverted");
+      return BinaryOperator::CreateSub(NotOp1, NotOp0);
+    }
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2208]: I strongly suspect that this fold is going to bite us now that we do the recursive walk. We can go back and forth between `(~X) - (~Y)` and `Y - X` as long as all of `X`, `Y`, `~X`, `~Y` are freely invertible.

Previously this would happen with constants (which this guards against), but I think now it can likely happen in more complex cases that root out at constants as well. I'm thinking something like `sub (select x, C1, C2), (select y, C3, C4)` might infinite loop, depending on whether the "binop of select" fold happens before this one or not. Or if not that, then maybe something like that with more operations in between.

Basically, for this transform it is important that we actually remove at least one instruction, not just "freely" invert in the sense that instruction count stays the same.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2208]: Ah, yeah thats a good point.

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 14]
```diff
@@ -233,58 +233,114 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
     // instruction.
-    bool IsSelect = match(V, m_Select(m_Value(), m_Value(A), m_Value(B))) &&
+    bool IsSelect = match(V, m_Select(m_Value(Cond), m_Value(A), m_Value(B))) &&
                     !shouldAvoidAbsorbingNotIntoSelect(*cast<SelectInst>(V));
     // Selects/min/max with invertible operands are freely invertible
-    if (IsSelect || match(V, m_MaxOrMin(m_Value(A), m_Value(B))))
-      return isFreeToInvert(A, A->hasOneUse(), Depth) &&
-             isFreeToInvert(B, B->hasOneUse(), Depth);
+    if (IsSelect || match(V, m_MaxOrMin(m_Value(A), m_Value(B)))) {
+      if (!getFreelyInverted(A, A->hasOneUse(), /*Builder*/ nullptr, Depth))
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:315]: Can do so elsewhere as well.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp (hunk 13 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2208]: Adding tests for all binops we support with lhs/rhs pattern that is always freely invertible.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 12 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4338]: Yup, dropping.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 11 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4392]: Was actually just operand order changes. Dropping `sinkNotIntoXor`.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 15]
```diff
@@ -2106,6 +2106,99 @@ Instruction *InstCombinerImpl::visitGEPOfGEP(GetElementPtrInst &GEP,
   return nullptr;
 }
 
+Value *InstCombiner::getFreelyInverted(Value *V, bool WillInvertAllUses,
+                                       BuilderTy *Builder, unsigned Depth) {
+  using namespace llvm::PatternMatch;
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2111]: Can probably drop the `using namespace` now.

File: llvm/test/Transforms/LoopVectorize/ARM/mve-selectandorcost.ll [hunk 16]
```diff
@@ -86,7 +86,7 @@ define float @test(ptr nocapture readonly %pA, ptr nocapture readonly %pB, i32 %
 ; CHECK-NEXT:    [[ACCUM_1]] = phi float [ [[ADD4]], [[IF_THEN]] ], [ [[ACCUM_017]], [[WHILE_BODY]] ]
 ; CHECK-NEXT:    [[DEC]] = add i32 [[BLOCKSIZE_ADDR_018]], -1
 ; CHECK-NEXT:    [[CMP_NOT:%.*]] = icmp eq i32 [[DEC]], 0
-; CHECK-NEXT:    br i1 [[CMP_NOT]], label [[WHILE_END]], label [[WHILE_BODY]], !llvm.loop [[LOOP2:![0-9]+]]
+; CHECK-NEXT:    br i1 [[CMP_NOT]], label [[WHILE_END]], label [[WHILE_BODY]], !llvm.loop [[LOOP3:![0-9]+]]
```
Comment (nikic) [inline llvm/test/Transforms/LoopVectorize/ARM/mve-selectandorcost.ll:89]: You can probably remove the diff in this test?

File: llvm/test/Transforms/InstCombine/free-inversion.ll [hunk 17]
```diff
@@ -0,0 +1,521 @@
...
+  %not_ba = xor i8 %ba, -1
+  ret i8 %not_ba
+}
+
+define i8 @sub_both_freely_invertable_always(i8 %x, i8 %y) {
+; CHECK-LABEL: @sub_both_freely_invertable_always(
+; CHECK-NEXT:    [[XX:%.*]] = xor i8 [[X:%.*]], 123
+; CHECK-NEXT:    [[TMP1:%.*]] = xor i8 [[Y:%.*]], -46
+; CHECK-NEXT:    [[YY_NEG:%.*]] = add i8 [[TMP1]], 1
+; CHECK-NEXT:    [[R:%.*]] = add i8 [[YY_NEG]], [[XX]]
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/free-inversion.ll:384]: Pre-existing issue: We're adding an extra instruction here. It looks like this bit in InstCombineNegator is responsible: https://github.com/llvm/llvm-project/blob/8748b3979ce414025825631dd77c27c58a4ed04b/llvm/lib/Transforms/InstCombine/InstCombineNegator.cpp#L445-L455 It's missing a IsTrulyNegation check. Will we run into an infinite loop here once we add it?

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/free-inversion.ll:384]: Yeah, you're right. Ill post patch for `IsTrulyNegatable` and add option for `MustConsumeInst` in `getFreelyInverted`.

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/free-inversion.ll:384]: See: https://github.com/llvm/llvm-project/pull/72767, will repost w/ `MustConsumeInst` flag once thats in.

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/free-inversion.ll]: Okay, update so that we ensure we save an inst.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 18]
```diff
@@ -2205,12 +2227,23 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
...
   // sub-of-sub pattern to avoid infinite looping.
-  if (isFreeToInvert(Op0, Op0->hasOneUse()) &&
-      isFreeToInvert(Op1, Op1->hasOneUse()) &&
-      !match(Op0, m_Sub(m_ImmConstant(), m_Value()))) {
-    Value *NotOp0 = Builder.CreateNot(Op0);
-    Value *NotOp1 = Builder.CreateNot(Op1);
-    return BinaryOperator::CreateSub(NotOp1, NotOp0);
+  {
+    auto m_AddSubInvertable = m_CombineOr(m_Add(m_Value(), m_ImmConstant()),
+                                          m_Sub(m_ImmConstant(), m_Value()));
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2232]: What is this for?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2232]: The `sub` is unnecessary (will drop). But we rely on this fold for simplifying things like `(A + C0) - (B + C1)`

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 19]
```diff
@@ -2205,12 +2223,22 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
...
+  {
+    // Need to ensure we can consume at least one of the `not` instructions,
+    // otherwise this can inf loop.
+    bool ConsumesOp0, ConsumesOp1;
+    if ((match(Op0, m_Add(m_Value(), m_ImmConstant())) &&
+         match(Op1, m_Add(m_Value(), m_ImmConstant()))) ||
+        (isFreeToInvert(Op0, Op0->hasOneUse(), ConsumesOp0) &&
+         isFreeToInvert(Op1, Op1->hasOneUse(), ConsumesOp1) &&
+         (ConsumesOp0 || ConsumesOp1) &&
+         !match(Op0, m_Sub(m_ImmConstant(), m_Value())))) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2235]: Is this still needed?

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 20]
```diff
@@ -2205,12 +2223,22 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
...
-      !match(Op0, m_Sub(m_ImmConstant(), m_Value()))) {
-    Value *NotOp0 = Builder.CreateNot(Op0);
-    Value *NotOp1 = Builder.CreateNot(Op1);
-    return BinaryOperator::CreateSub(NotOp1, NotOp0);
+  {
+    // Need to ensure we can consume at least one of the `not` instructions,
+    // otherwise this can inf loop.
+    bool ConsumesOp0, ConsumesOp1;
+    if ((match(Op0, m_Add(m_Value(), m_ImmConstant())) &&
+         match(Op1, m_Add(m_Value(), m_ImmConstant()))) ||
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2231]: Isn't this going to do the wrong thing (or maybe assert?) if Op0 or Op1 are adds with extra uses?

Do I understand correctly that this folds `(X + C1) - (Y + C2)` into `(~C1 - X) - (~C2 - Y)` and then some other fold is going to fold that into `(X + Y) + (C1 - C2)`? If so, that seems super convoluted, should we just have an explicit fold to do that instead?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2231]: Yeah, thats probably cleaner. Will do.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp (hunk 19 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2235]: Nope, dropping.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp (hunk 20 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2231]: Done.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp (hunk 19 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2235]: Done.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 21]
```diff
@@ -2224,12 +2224,17 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
   // (~X) - (~Y) --> Y - X
   // This is placed after the other reassociations and explicitly excludes a
   // sub-of-sub pattern to avoid infinite looping.
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2249]: The comment above is outdated / redundant with the one below.

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 22]
```diff
@@ -233,66 +233,38 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
                                                 PatternMatch::m_Value()));
   }
 
+  /// Return nonnull value if V is free to invert (with condition) regarding
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:236]: I'm not sure what the `(with condition)` bit here is supposed to mean.

File: llvm/include/llvm/Transforms/InstCombine/InstCombiner.h [hunk 23]
```diff
@@ -233,66 +233,38 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
...
+  /// WillInvertAllUses.
+  /// If Builder is nonnull, it will return a simplified ~V
```
Comment (nikic) [inline llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:238]: nit: period at end of sentence.

**Review by nikic — APPROVED** [review]
LGTM with the last commit removed (the test coverage is quite insufficient around m_AddSubInvertable, and I don't particularly want this fold anyway).

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp (hunk 21 above)
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2249]: Well, at least the part about the "sub-of-sub pattern".

**Comment by goldsteinn** [comment]
> LGTM with the last commit removed (the test coverage is quite insufficient around m_AddSubInvertable, and I don't particularly want this fold anyway).

I'd be happy to improve the coverage. Why are you opposed to the commit as a whole?

**Comment by nikic** [comment]
> > LGTM with the last commit removed (the test coverage is quite insufficient around m_AddSubInvertable, and I don't particularly want this fold anyway).
> 
> I'd be happy to improve the coverage. Why are you opposed to the commit as a whole? Also do you mean removed entirely, or is changing is to `ConsumeLHS && ConsumeRHS` okay?

It looks like a fold with very low value but very high risk. Limiting it to `ConsumeLHS && ConsumeRHS` (with `&&`, not `||`, and no special add/sub cases) would probably be acceptable.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp (hunk 21 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2249]: Although the whole thing is deprecated as ensuring we consume a `not` is what precludes inf loop.

**Comment by goldsteinn** [comment]
> > > LGTM with the last commit removed (the test coverage is quite insufficient around m_AddSubInvertable, and I don't particularly want this fold anyway).
> > 
> > 
> > I'd be happy to improve the coverage. Why are you opposed to the commit as a whole? Also do you mean removed entirely, or is changing is to `ConsumeLHS && ConsumeRHS` okay?
> 
> It looks like a fold with very low value but very high risk. Limiting it to `ConsumeLHS && ConsumeRHS` (with `&&`, not `||`, and no special add/sub cases) would probably be acceptable.
Okay, done (added b.c noticed after changing the logic for `~Y - ~X` had a regression).

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 24]
```diff
@@ -1662,6 +1662,24 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
...
     return BinaryOperator::CreateWithCopiedFlags(Instruction::Add, A, B, &I);
 
+  // (~X) + (~Y) --> (-2 - Y) - X
+  {
+    // To ensure we can save instructions we need to ensure that we either
+    // consume both LHS/RHS (i.e they have a `not`).
+    bool ConsumesLHS, ConsumesRHS;
+    if (isFreeToInvert(LHS, LHS->hasOneUse(), ConsumesLHS) &&
+        isFreeToInvert(RHS, RHS->hasOneUse(), ConsumesRHS) && ConsumesLHS &&
+        ConsumesRHS) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1672]:
```suggestion
    if (isFreeToInvert(LHS, LHS->hasOneUse(), ConsumesLHS) && ConsumeLHS &&
        isFreeToInvert(RHS, RHS->hasOneUse(), ConsumesRHS) && ConsumesRHS) {
```
Can short-circuit the second call.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 25]
```diff
@@ -1662,6 +1662,24 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
                                     m_c_UMin(m_Deferred(A), m_Deferred(B))))))
     return BinaryOperator::CreateWithCopiedFlags(Instruction::Add, A, B, &I);
 
+  // (~X) + (~Y) --> (-2 - Y) - X
+  {
+    // To ensure we can save instructions we need to ensure that we either
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1667]:
```suggestion
    // To ensure we can save instructions we need to ensure that we
```

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 26]
```diff
@@ -1662,6 +1662,24 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
...
+    if (isFreeToInvert(LHS, LHS->hasOneUse(), ConsumesLHS) &&
+        isFreeToInvert(RHS, RHS->hasOneUse(), ConsumesRHS) && ConsumesLHS &&
+        ConsumesRHS) {
+      Value *NotLHS = getFreelyInverted(LHS, LHS->hasOneUse(), &Builder);
+      Value *NotRHS = getFreelyInverted(RHS, RHS->hasOneUse(), &Builder);
+      assert(NotLHS != nullptr && NotRHS != nullptr &&
+             "isFreeToInvert desynced with getFreelyInverted");
+      Value *RHSMinus2 =
+          Builder.CreateSub(ConstantInt::get(RHS->getType(), -2), NotRHS);
+      return BinaryOperator::CreateSub(RHSMinus2, NotLHS);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1679]: I believe InstCombine's canonical form for this is `-2 - (X + Y)` rather than `(-2 - X) - Y`.

**Review by nikic — APPROVED** [review]
LGTM
